client/src/wayland_obj/output.rs

```diff
@@ -82,52 +91,31 @@ impl wl_output::EventHandler for WlOutputEventHandler {
     #[allow(unused)]
     fn mode(&mut self, object: WlOutput, flags: wl_output::Mode, width: i32, height: i32, refresh: i32) {
         unwrap_state(object.as_ref()).borrow_mut().resolution = (width as u32, height as u32);
-        let geometry = Area {
-            origin: Origin { x: 0, y: 0 },
-            size: Size {
-                width: width as u32,
-                height: height as u32
-            }
-        };
-        LUA.with(|lua| {
-            lua.borrow().context(|ctx| {
-                if let Ok(mut screen) = screen::get_screen(ctx, Output { output: object }) {
-                    screen
-                        .set_geometry(ctx, geometry)
-                        .expect("could not set geometry");
-                    screen
-                        .set_workarea(ctx, geometry)
-                        .expect("could not set workarea ");
-                }
-            });
-        });
     }
 
     #[allow(unused)]
     fn done(&mut self, object: WlOutput) {
-        // TODO We may not always want to add a new screen
-        // see how awesome does it and fix this.
-        LUA.with(|lua| {
-            lua.borrow().context(|ctx| {
-                let mut screen = Screen::new(ctx).expect("Could not allocate new screen");
-                let output = Output { output: object };
-                screen
-                    .init_screens(output.clone(), vec![output])
-                    .expect("Could not initilize new output with a screen");
-                screen::add_screen(ctx, screen).expect("Could not add screen to the list of screens");
-            });
-        });
+        self.handler.output_changed(Output { output: object });
     }
 
     #[allow(unused)]
     fn scale(&mut self, object: WlOutput, factor: i32) {
         // TODO
+        if factor != 1 {
+            unimplemented!("output scaling");
+        }
     }
 }
 
 impl fmt::Debug for Output {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{:?}", self.output.as_ref().c_ptr())
+        write!(f, "{:?}", self.name())
+    }
+}
+
+impl AsRef<WlOutput> for Output {
+    fn as_ref(&self) -> &WlOutput {
```

Comment:
General rust question: Use `#[inline]` or not?

Reply:
Nah, trust the optimizer to do it (or not). There's too much voodoo with `#[inline]`. If this really is a problem then we would first:

1) Profile, and ensure that a callstack really is too much overhead (doubtful)
2) Enable LTO, which would probably enable automatic inlining
3) Add `#[inline]`, _not_ `#[inline(always)]` which could mess with debugging IIRC.

Generally though, don't worry about it.